File: WorkcraftCore/src/main/java/org/workcraft/gui/propertyeditor/EditableProperty.java
/*
*
* Copyright 2008,2009 Newcastle University
*
* This file is part of Workcraft.
* 
* Workcraft is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* Workcraft is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with Workcraft.  If not, see <http://www.gnu.org/licenses/>.
*
*/

package org.workcraft.gui.propertyeditor;

import java.awt.BorderLayout;
import java.awt.Component;

import javax.swing.JPanel;

import org.workcraft.dependencymanager.advanced.core.Expression;
import org.workcraft.dependencymanager.advanced.core.Expressions;
import org.workcraft.dependencymanager.advanced.user.ModifiableExpression;
import org.workcraft.util.Action;
import org.workcraft.util.Function;
import org.workcraft.util.Nothing;

public interface EditableProperty {
	class Util {
		public static <T> EditableProperty create(final String propertyName, final GenericEditorProvider<T> editor, final RendererProvider<T> renderer, final ModifiableExpression<T> property) {
			
			return new EditableProperty(){
				@Override
				public ReactiveComponent renderer(Expression<Boolean> isSelected, Expression<Boolean> hasFocus) {
					final JPanel panel = new JPanel();
					return new ReactiveComponent() {
						@Override
						public Expression<? extends Nothing> updateExpression() {
							return Expressions.fmap(new Function<T, Nothing>(){
								@Override
								public Nothing apply(T value) {
									panel.removeAll();
									panel.setLayout(new BorderLayout());
									panel.add(renderer.createRenderer(value), BorderLayout.CENTER);
									return Nothing.VALUE;
								}
							}, property);
						}
						@Override
						public Component component() {
							return panel;
						}
					};
				}

				@Override
				public Expression<? extends EditorProvider> editorMaker() {
					return Expressions.fmap(new Function<T, EditorProvider>(){

						@Override
						public EditorProvider apply(final T value) {
							return new EditorProvider() {
								
								@Override
								public SimpleCellEditor getEditor(final Action close) {
									return new SimpleCellEditor() {

										GenericCellEditor<T> ge = editor.createEditor(value, new Action(){

											@Override
											public void run() {
												commit();
												close.run();
											}
										}, close);
										
										@Override
										public Component getComponent() {
											return ge.component();
										}
										
										@Override
										public void commit() {
											property.setValue(ge.getValue());
										}
									};
								}
							};
						}
						
					}, property);
				}

				@Override
				public String name() {
					return propertyName;
				}
			};
		}
	}
	public String name();
	public ReactiveComponent renderer(Expression<Boolean> isSelected, Expression<Boolean> hasFocus);
	public Expression<? extends EditorProvider> editorMaker();
}

File: Gui/src/main/scala/org/workcraft/gui/propertyeditor/EditableProperty.scala
package org.workcraft.gui.propertyeditor

import java.awt.BorderLayout
import java.awt.Component
import javax.swing.JPanel
import org.workcraft.scala.Expressions._
import org.workcraft.scala.effects.IO
import org.workcraft.scala.effects.IO._
import scalaz._
import Scalaz._
import javax.swing.JComponent

trait EditableProperty {
  val name: String
  def renderer(isSelected: Boolean, hasFocus: Boolean): JComponent
  def createEditor(accept: IO[Unit], cancel: IO[Unit]): SimpleCellEditor
}

object EditableProperty {
  def apply[T](propertyName: String, editor: GenericEditorProvider[T], _renderer: RendererProvider[T], value: T, _commit: T => IO[Unit]): EditableProperty = {
    withValidation(propertyName, editor, _renderer, value, (t: T) => ioPure.pure { try {_commit(t).unsafePerformIO; None } catch { case e: Throwable => Some(e.toString)}})
  }

  def withValidation[T](propertyName: String, editor: GenericEditorProvider[T], _renderer: RendererProvider[T], value: T, _commit: T => IO[Option[String]]): EditableProperty =
    new EditableProperty {
      override def renderer(isSelected: Boolean, hasFocus: Boolean) = {
        val panel = new JPanel
        panel.setLayout(new BorderLayout)
        panel.add(_renderer.createRenderer(value), BorderLayout.CENTER)
        panel
      }

      override def createEditor(accept: IO[Unit], cancel: IO[Unit]) = new SimpleCellEditor {
        val ge: GenericCellEditor[T] = editor.createEditor(value, accept, cancel)
        
        override def getComponent = ge.component
        override def commit = _commit( ge.getValue )
      }

      override val name = propertyName
    }
  
  def apply[T](propertyName: String, editor: GenericEditorProvider[T], renderer: RendererProvider[T], mewv : ModifiableExpressionWithValidation[T, String]) : Expression[EditableProperty] = {
    mewv.expr.map(value => {
      withValidation(propertyName, editor, renderer, value, (t:T) => mewv.set(t))
    })
  }
}
